source/common/network/connection_impl.cc

```diff
@@ -34,6 +34,16 @@ Address::InstanceConstSharedPtr getNullLocalAddress(const Address::Instance& add
 }
 } // namespace
 
+int ConnectionImplUtility::createSocket(Address::InstanceConstSharedPtr address,
+                                        Optional<Address::InstanceConstSharedPtr> source_address) {
+  int fd = address->socket(Address::SocketType::Stream);
+  if (fd >= 0 && source_address.valid()) {
+    int rc = source_address.value()->bind(fd);
+    ASSERT(rc >= 0);
```

Comment:
Ok.  How about I move the RELEASE_ASSERT in the connection to the socket creation, and then in the connection treat -1  as an immediate disconnect and increment some stat so folks can see why all the connections to a given [statscope] are failing?